---
title: "Linear Model Challenge"
subtitle: "Starting kit"
author: "Florent Chuffart & Magali Richard"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: false
---

```{r label="header", echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
# genes = readRDS("~/projects/genes/bed_grch38_epimeddb.rds")
# dim(genes)
```


# Phase 1

## Data

The challenge provides the data.frame `e` of non tumoral tissues described by genes expression values, histological and clinical attributes.


**TODO, TP M1 SSD: data descriptive statistics**

```{r results="verbatim", echo=TRUE}
e = readRDS(file="chlg1_modlin_e.rds")
head(e)
dim(e)
```


## LRRC1 gene expression

Some LRRC1 expression values are missing.

**The goal of the challenge** is to illustrate linear model properpties by predicting LRRC1 genes expression missing values according histological and clinical attributes.

LRRC1 contributes to HCC development, and may be a potential target for therapeutic intervention in this disease [Li2013].

According to the litterature, expression of LRRC1 is broad in colon, thyroid and some other tissues [NCBI/GENE]. 


```{r}
table(is.na(e$LRRC1))
```

## Method

We dispose sample from many tissues, the idea is to evaluate if *tissue* could be a good predictor for *LRRC1 expression*.


After checking model hypothesis (**TODO, TP M1 SSD**), we use ANOVA to predict missing values.


```{r}
layout(1, respect=TRUE)
boxplot(LRRC1~tissue, e, las=2, main="LRRC1~tissue")
```
```{r results="verbatim", echo=TRUE}
sort(table(e$tissue))
# ANOVA
m = lm(LRRC1~tissue, e)
## model hypothesis ################## (**TODO, TP M1 SSD**) ##################
## model results
summary(m)
anova(m)
# prediction
e$LRRC1_pred = e$LRRC1
idx = is.na(e$LRRC1)
e[idx,]$LRRC1_pred = predict(m, e[idx,])
```

```{r}
layout(matrix(1:2,1), respect=TRUE)
boxplot(LRRC1~tissue, e, las=2, main="LRRC1~tissue in training set")
boxplot(LRRC1_pred~tissue, e[idx,], las=2, main="prediction of LRRC1~tissue")
```


## Results

Then, we export results for submission.


The metric used to evaluate prediction is the  **mean squared error** (MSE)

$$MSE = \frac{1}{n}\sum_{i=1}^n{(Y_i-\widehat{Y}_i )^2}$$



The prediction evaluation is done using the script *scoring_program/scoring.r* (provided).

```{r echo=TRUE, results="verbatim"}
# Export results
e_LRRC1 = e$LRRC1_pred 
names(e_LRRC1) = rownames(e)
saveRDS(e_LRRC1, "chlg1_modlin_res.rds")
zip_filename = paste0("chlg1_modlin_res_", format(Sys.time(), format="%m_%d_%Y_%s"), ".zip")
zip(zip_filename, "chlg1_modlin_res.rds")
print(zip_filename)
```

```{r, code = readLines("scoring_program/scoring.r"), results="verbatim"}
```



## Submission

Produced results will be submitted to the "Linear Model Challenge" in the *Participate* tab.

https://competitions.codalab.org/competitions/21774#participate

How to submit results?
Produced results will be submitted to the “Linear Model Challenge” in the Participate tab.



## To go further

**TODO, TP M1 SSD**

1. Does linear regression model $LRRC1 \sim age$ provide better results?

```{r}
layout(matrix(1:2,1), respect=TRUE)
plot(e$age, e$LRRC1, main="LRRC1~age")
boxplot(e$age~e$tissue, las=2)
```


Not linear...
 
Desequiliobrium between low expressiuon of LRRC1 and age. Desiquilibrium between tissue and age? 

Confondant factor between thyroid and age...

```{r echo=TRUE, results="verbatim"}
m = lm(LRRC1~age, e)
summary(m)
```

2. Does ANOVA model $LRRC1 \sim sex$ provide better results?

```{r}
layout(1, respect=TRUE)
boxplot(e$LRRC1~e$sex, main="LRRC1~sex")
```

```{r echo=TRUE, results="verbatim"}
m = lm(LRRC1~sex, e)
summary(m)
```

Not evident...


3. Does ANOVA $LRRC1 \sim tissue+sex$ provide better results?

```{r}
layout(1, respect=TRUE)
boxplot(LRRC1~sex+tissue, e, main="LRRC1~tissue+sex", las=2)
```


correlated effect: not the same influence (up/down) of sex according to tissue.

```{r echo=TRUE, results="verbatim"}
m = lm(LRRC1~tissue + sex, e)
summary(m)


# Export results
e_LRRC1 = e$LRRC1_pred 
names(e_LRRC1) = rownames(e)
saveRDS(e_LRRC1, "chlg1_modlin_res.rds")
zip_filename = paste0("chlg1_modlin_res_", format(Sys.time(), format="%m_%d_%Y_%s"), ".zip")
zip(zip_filename, "chlg1_modlin_res.rds")
print(zip_filename)
```

```{r}
# "lm vs. oav"  "+ vs. *"
m1 = lm(LRRC1~tissue+sex, e)
e$LRRC1_pred1 = e$LRRC1
idx = is.na(e$LRRC1)
e[idx,]$LRRC1_pred1 = predict(m1, e[idx,])

m2 = lm(LRRC1~tissue*sex, e)
e$LRRC1_pred2 = e$LRRC1
idx = is.na(e$LRRC1)
e[idx,]$LRRC1_pred2 = predict(m2, e[idx,])

e$LRRC1_pred2 == e$LRRC1_pred1

m3 = aov(LRRC1~tissue+sex, e)
e$LRRC1_pred3 = e$LRRC1
idx = is.na(e$LRRC1)
e[idx,]$LRRC1_pred3 = predict(m3, e[idx,])

m4 = aov(LRRC1~tissue*sex, e)
e$LRRC1_pred4 = e$LRRC1
idx = is.na(e$LRRC1)
e[idx,]$LRRC1_pred4 = predict(m4, e[idx,])

e$LRRC1_pred3 == e$LRRC1_pred4

e$LRRC1_pred1 == e$LRRC1_pred3
e$LRRC1_pred2 == e$LRRC1_pred4
```


4. The challenge provides the gene expression matrix  `d` for the  previously described non tumoral tissues.
Does exist a gene *g* for which linear regression model $LRRC1 \sim g$ provide better results?

5. What is your best model according to provided data?


```{r results="verbatim", echo=TRUE}
d = readRDS(file="chlg1_modlin_d.rds")
dim(d)
e = cbind(e, t(d[,rownames(e)]))
dim(e)
# linear regression according an other gene
m = lm(LRRC1~ACAP2, e)
summary(m)
m = lm(LRRC1~LRRC38, e)
summary(m)
# selecting genes according to correlation with LRRC1
idx = !is.na(e$LRRC1)
cors = sapply(rownames(d),function(gs) {
  cor(e[idx,"LRRC1"], e[idx,gs])  
})
head(sort(cors, decreasing=TRUE))
tail(sort(cors, decreasing=TRUE))
head(sort(abs(cors)))
m = lm(LRRC1~SPINT2, e)
summary(m)
layout(matrix(1:2,1), respect=TRUE)
plot(e$SPINT2, e$LRRC1, xlab="SPINT2", ylab="LRRC1", )
plot(e$CD40, e$LRRC1, xlab="CD40", ylab="LRRC1", )
```


















# References 


[NCBI/GENE] https://www.ncbi.nlm.nih.gov/gene/55227

[Li2013] Aberrant upregulation of LRRC1 contributes to human hepatocellular carcinoma.
Li Y, et al. Mol Biol Rep, 2013 Jul. PMID 23645086

[TCGA] Data are coming from the TCGA (The cancer genome atlas). 
They were extracted from severla cohorts of patients. 
Here are TCGA Study Abbreviations corresponding to each cancer. 
https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations

[Codalab] https://competitions.codalab.org


**Supplementals**

`data` corresponds to gene expression data for each patients.

`pf` corresponds to informations related to each gene.

```{r eval=FALSE, echo=FALSE}
data    = readRDS(file="chlg1_modlin_d.rds")
head(data)
dim(data)
pf      = readRDS(file="chlg1_modlin_p.rds")
head(pf)
dim(pf)
```


# Informations


**How to participate?**

The challenge is hosted on the codalab platform:

https://competitions.codalab.org/competitions/21774?secret_key=a58ff8ad-c04c-4157-80fd-ac07c3473a68

To participate, you need to:

1- create an account on the codalab platform

2- subscribe to the **Linear Model Challenge**


**How to start?**

First, download the **starting kit** in the *get_starting_kit* section.

Then, go to the *strating-k* directory.

Finally, launch in R the following command:

```
rmarkdown::render("starting_kit.Rmd")
```

It procudes the file "starting_kit.html" that explains how to analyse data and produce results.

**How to submit results?**

Produced results will be submitted to the "Linear Model Challenge" in the *Participate* tab.
  







```{r eval=FALSE}
####  PCA Guttman effect


pca = prcomp(t(data), scale=FALSE)
v = pca$sdev * pca$sdev
p = v / sum(v) * 100


col = as.numeric(as.factor(exp_grp[idx_nt,]$tissue_group_level1))
layout(matrix(1:2,1), respect=TRUE)
barplot(p[1:10], ylab="% of var")

i = 1
j = 2
plot(pca$x[,i], pca$x[,j], 
  xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), 
  ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"),
  col=col
)
legend("topright", pch=16, col=1:length(unique(exp_grp[idx_nt,]$tissue_group_level1)), levels(as.factor(exp_grp[idx_nt,]$tissue_group_level1)))


for (i in c(1,3,5,7)) {
  j = i+1
  plot(pca$x[,i], pca$x[,j], 
    xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), 
    ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"),
    col=col
  )  
}

```














# Session Information

```{r, results="verbatim"}
sessionInfo()
```


