---
title: "Linear Model Challenge"
subtitle: "Starting kit"
author: "Florent Chuffart & Magali Richard"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: false
---

```{r label="header", echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
# genes = readRDS("~/projects/genes/bed_grch38_epimeddb.rds")
# dim(genes)
```


# Overview of the challenge



The challenge provides the data.frame `e` of non tumoral tissues described by genes expression values, histological and clinical attributes.


```{r results="verbatim", echo=TRUE}
e = readRDS(file="chlg1_modlin_e.rds")
head(e)
dim(e)
```

Some LRRC1 expression values are missing.

```{r}
table(is.na(e$LRRC1))
```

**The goal of the challenge** is to illustrate linear model properpties by predicting LRRC1 genes expression missing values according histological and clinical attributes.

According to the litterature, expression of LRRC1 is broad in colon, thyroid and some other tissues.

https://www.ncbi.nlm.nih.gov/gene/55227

```{r results="verbatim", echo=TRUE}
sort(table(e$tissue))
layout(matrix(1:2,1), respect=TRUE)
boxplot(LRRC1~tissue, e, las=2, main="LRRC1~tissue")
```

After checking model hypothesis (**TODO, TP M1 SSD**), we use ANOVA to predict missing values.

```{r results="verbatim", echo=TRUE}
# ANOVA
m = lm(LRRC1~tissue, e)
## model hypothesis ################## (**TODO, TP M1 SSD**) ##################
## model results
summary(m)
anova(m)
# prediction
e$LRRC1_pred = e$LRRC1
idx = is.na(e$LRRC1)
e[idx,]$LRRC1_pred = predict(m, e[idx,])
```

```{r}
layout(matrix(1:2,1), respect=TRUE)
boxplot(LRRC1~tissue, e, las=2, main="LRRC1~tissue in training set")
boxplot(LRRC1_pred~tissue, e[idx,], las=2, main="prediction of LRRC1~tissue")
```




Then, we export results for submission.

```{r echo=TRUE, results="verbatim"}
# Export results
e_LRRC1 = e$LRRC1_pred 
names(e_LRRC1) = rownames(e)
saveRDS(e_LRRC1, "chlg1_modlin_res.rds")
zip_filename = paste0("chlg1_modlin_res_", format(Sys.time(), format="%B_%d_%Y_%s"), ".zip")
zip(zip_filename, "chlg1_modlin_res.rds")
# file.remove("chlg1_modlin_res.rds")
```


The metric used to evaluate prediction is the  **mean squared error** (MSE)

$$MSE = \frac{1}{n}\sum_{i=1}^n{(Y_i-\widehat{Y}_i )^2}$$



The prediction evaluation is done using the script *scoring_program/scoring.r* (provided).

```{r}
input = output = "./"
```
```{r, code = readLines("scoring_program/scoring.r"), results="verbatim"}
```



  


```{r eval=FALSE}
table(e$sex)
boxplot(LRRC1~sex, e, las=2, main="LRRC1~sex")

plot(e$age, e$LRRC1, main="LRRC1~age")

table(is.na(e$LRRC1))
hist(e$LRRC1)

# m = lm(LRRC1~tissue, e)
# s = summary(m)
# s$r.squared
```











# Evaluation: codalab platform 

## How to participate?

The challenge is hosted on the codalab platform:

https://competitions.codalab.org/competitions/21774?secret_key=a58ff8ad-c04c-4157-80fd-ac07c3473a68

To participate, you need to:

1- create an account on the codalab platform

2- subscribe to the **Linear Model Challenge**


## How to start?

First, download the **starting kit** in the *get_starting_kit* section.

Then, go to the *strating-k* directory.

Finally, launch in R the following command:

```
rmarkdown::render("starting_kit.Rmd")
```

It procudes the file "starting_kit.html" that explains how to analyse data and produce results.

## How to submit results?

Produced results will be submitted to the "Linear Model Challenge" in the *Participate* tab.
  


















# References & Supplementals



Data are coming from the TCGA (The cancer genome atlas). They were extracted from severla cohorts of patients. Here are TCGA Study Abbreviations corresponding to each cancer:

https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations





```{r eval=FALSE, echo=FALSE}
# `data` corresponds to gene expression data for each pations patients (9466 genes).
#
# `pf` corresponds to informations related to each gene.

data    = readRDS(file="chlg1_modlin_d.rds")
head(data)
dim(data)
pf      = readRDS(file="chlg1_modlin_p.rds")
head(pf)
dim(pf)

```



```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="hide")
require(survival)
scurve = function (ss, v, colors = c("deepskyblue", "black", "red"), main = "Survival", 
    legend, nb_sign = 3, legend_place = "topright", PLOT_LEGEND = TRUE, 
    cex.leg = 1, censoring, ...) 
{
    idx = !is.na(ss)
    ss = ss[idx]
    if (missing(v)) {
        v = rep(1, sum(idx))
    }
    else {
        v = v[idx]
    }
    if (!missing(censoring)) {
        date = ss[, 1]
        state = ss[, 2]
        idx = !is.na(date) & date > censoring
        date[idx] = censoring
        state[idx] = 0
        ss = survival::Surv(date, state)
    }
    if (length(unique(v)) == 1) {
        cox_results = 1
    }
    else {
        cox_results = coxres(ss, as.character(v))
    }
    pv = cox_results[1]
    if (pv < 1e-100) {
        pvt = "<1e-100"
    }
    else {
        pvt = format(pv, digits = 3, scientific = TRUE)
    }
    sf = survival::survfit(ss ~ v)
    levels = length(na.omit(unique(v)))
    col = colorRampPalette(colors)(levels)
    main = paste(main, " p_cox=", signif(as.numeric(pvt), nb_sign), 
        sep = "")
    plot(sf, col = col, main = main, ...)
    tab = table(v)
    tab = tab[tab != 0]
    if (PLOT_LEGEND) {
        if (missing(legend)) {
            if ("breaks" %in% names(attributes(v))) {
                b = signif(attr(v, "breaks"), 3)
                legend = paste("[", b[1:(length(b) - 1)], ",", 
                  b[2:length(b)], c(rep("[", length(b) - 2), 
                    "]"), sep = "")
            }
            else {
                legend = names(tab)
            }
        }
        legend = paste(legend, " (", tab, ")", sep = "")
        legend(legend_place, legend = legend, col = col, pch = 3, 
            lty = 1, cex = cex.leg)
    }
    return(cox_results)
}

coxres = function (ss, v) 
{
    f = suppressWarnings(survival::coxph(ss ~ v))
    sf = summary(f)
    pvcox = sf$logtest[3]
    tf = survival::cox.zph(f)
    tf = tf$table
    pvhz = tf[dim(tf)[1], 3]
    hr = sf$conf.int[1, 1]
    hrlb = sf$conf.int[1, 3]
    hrub = sf$conf.int[1, 4]
    res = c(pvcox, pvhz, hrlb, hr, hrub)
    names(res) = c("pvcox", "pvhz", "hrlb", "hr", "hrub")
    return(res)
}
```


```{r eval=FALSE}
## Tumoral and non tumoral samples 
idx_nt = rownames(exp_grp)[exp_grp$tissue_status == "normal"] 
idx_tu = rownames(exp_grp)[exp_grp$tissue_status == "tumoral"] 
```

```{r eval=FALSE}
idx_training = rownames(exp_grp)[exp_grp$sets == "training"] 
length(idx_training)
idx_chlg1 = rownames(exp_grp)[exp_grp$sets == "chlg1"]
length(idx_chlg1)
idx_challenge2 = rownames(exp_grp)[exp_grp$sets == "challenge2"] 
length(idx_challenge2)
```



##  PCA Non tumoral tissues

Guttman effect

```{r eval=FALSE}
pca = prcomp(t(data[,idx_nt]), scale=FALSE)
v = pca$sdev * pca$sdev
p = v / sum(v) * 100


col = as.numeric(as.factor(exp_grp[idx_nt,]$tissue_group_level1))
layout(matrix(1:2,1), respect=TRUE)
barplot(p[1:10], ylab="% of var")

i = 1
j = 2
plot(pca$x[,i], pca$x[,j], 
  xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), 
  ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"),
  col=col
)
legend("topright", pch=16, col=1:length(unique(exp_grp[idx_nt,]$tissue_group_level1)), levels(as.factor(exp_grp[idx_nt,]$tissue_group_level1)))


for (i in c(1,3,5,7)) {
  j = i+1
  plot(pca$x[,i], pca$x[,j], 
    xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), 
    ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"),
    col=col
  )  
}

```


```{r eval=FALSE}
## Tumoral tissues

pca = prcomp(t(data[,idx_tu]), scale=FALSE)
v = pca$sdev * pca$sdev
p = v / sum(v) * 100


col = as.numeric(as.factor(exp_grp[rownames(pca$x),]$main_gse_number))
layout(matrix(1:2,1), respect=TRUE)
barplot(p[1:10], ylab="% of var")

i = 1
j = 2
plot(pca$x[,i], pca$x[,j], 
  xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), 
  ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"),
  col=col
)
legend("topright", pch=16, col=1:length(unique(exp_grp[rownames(pca$x),]$main_gse_number)), levels(as.factor(exp_grp[rownames(pca$x),]$main_gse_number)))


# ```


# Survival

We plot survival according to: 

   - cancer type
  - sex
  - cancer stage 
  - level of expression of ATAD2 gene

# ```{r label="survival"}
layout(matrix(1:2, 1), respect=TRUE)
scurve(exp_grp$os, exp_grp$main_gse_number, main="cancer")
scurve(exp_grp$os, exp_grp$sex, main="sex")
scurve(exp_grp$os, exp_grp$tnm_stage, main="stage")
scurve(exp_grp$os, data["ATAD2",rownames(exp_grp)] > 10.5, main="ATAD2 over expression")
```



## Challenge 2

```{r eval=FALSE}
# Define groups
idx = rownames(exp_grp)[exp_grp$sets == "challenge2"]
grps = data["ATAD2",idx] < 10.5
names(grps) = idx

# Export grps 
saveRDS(grps, "grps.rds")
zip("grps_chal2.zip", "grps.rds")
```

# Session Information

```{r, results="verbatim"}
sessionInfo()
```


